streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -16,78 +16,408 @@
+     * to make assignment decisions with.
+     * This includes ensuring that every client has a known rack id, and that every topic partition for
+     * every logical task that needs to be assigned also has known rack ids.
+     * If a logical task has no source topic partitions, it will return false.
+     * If standby tasks are configured, but a logical task has no changelog topic partitions, it will return false.
      *
-     * @return a new map containing an assignment that replicates exactly the previous assignment reported
-     *         in the applicationState
+     * @return whether rack-aware assignment decisions can be made for this application.
      */
-    public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(
-        final ApplicationState applicationState
-    ) {
-        throw new UnsupportedOperationException("Not Implemented.");
+    private static boolean hasValidRackInformation(final ApplicationState applicationState) {
+        for (final KafkaStreamsState state : applicationState.kafkaStreamsStates(false).values()) {
+            if (!hasValidRackInformation(state)) {
+                return false;
+            }
+        }
+
+        for (final TaskInfo task : applicationState.allTasks()) {
+            if (!hasValidRackInformation(task)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean hasValidRackInformation(final KafkaStreamsState state) {
+        if (!state.rackId().isPresent()) {
+            LOG.error("Client " + state.processId() + " doesn't have rack configured.");
+            return false;
+        }
+        return true;
+    }
+
+    private static boolean hasValidRackInformation(final TaskInfo task) {
+        for (final TaskTopicPartition topicPartition : task.topicPartitions()) {
+            final Optional<Set<String>> racks = topicPartition.rackIds();
+            if (!racks.isPresent() || racks.get().isEmpty()) {
+                LOG.error("Topic partition {} for task {} does not have racks configured.", topicPartition, task.id());
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * This function returns a copy of the old collection of {@code KafkaStreamsAssignment} with tasks
+     * moved according to the {@param reassigned} tasks and {@param unassigned} tasks.
+     *
+     * @param kafkaStreamsAssignments the collection to start from when moving tasks from process to process
+     * @param reassigned              the map from process id to tasks that this client has newly been assigned
+     * @param unassigned              the map from process id to tasks that this client has newly been unassigned
+     *
+     * @return the new mapping from processId to {@code KafkaStreamsAssignment}.
+     */
+    private static Map<ProcessId, KafkaStreamsAssignment> processTaskMoves(final Collection<KafkaStreamsAssignment> kafkaStreamsAssignments,
```

Comment:
Alright I went ahead and tried this out for myself, I think it helps a lot to simplify things in this PR so it makes sense to incorporate here rather than doing it as an entirely separate PR. Lmk your thoughts as always: https://github.com/apourchet/kafka/pull/1